features/mbedtls/targets/TARGET_NUVOTON/TARGET_M480/aes/aes_alt.c

```diff
@@ -21,298 +21,213 @@
-            break;
-        default : return( MBEDTLS_ERR_AES_INVALID_KEY_LENGTH );
-    }
-
-
 
-	// key swap
-		for( i = 0; i < ( keybits >> 5 ); i++ )
-		{
-						ctx->buf[i] = (*(key+i*4) << 24) |
-													(*(key+1+i*4) << 16) |
-													(*(key+2+i*4) << 8) |
-													(*(key+3+i*4) );
-		}
-    AES_SetKey(ctx->channel, ctx->buf, ctx->keySize);
+    switch( keybits ) {
+    case 128:
+        ctx->keySize = AES_KEY_SIZE_128;
+        break;
+    case 192:
+        ctx->keySize = AES_KEY_SIZE_192;
+        break;
+    case 256:
+        ctx->keySize = AES_KEY_SIZE_256;
+        break;
+    default :
+        return( MBEDTLS_ERR_AES_INVALID_KEY_LENGTH );
+    }
 
+    /* Fetch key byte data in big-endian */
+    for( i = 0; i < ( keybits >> 5 ); i++ ) {
+        ctx->keys[i] = nu_get32_be(key + i * 4);
+    }
 
     return( 0 );
 }
-#endif /* MBEDTLS_AES_SETKEY_ENC_ALT */
 
 /*
  * AES key schedule (decryption)
  */
-#if defined(MBEDTLS_AES_SETKEY_DEC_ALT)
 int mbedtls_aes_setkey_dec( mbedtls_aes_context *ctx, const unsigned char *key,
-                    unsigned int keybits )
+                            unsigned int keybits )
 {
     int ret;
-	
-	  mbedtls_trace("=== %s keybits[%d]\r\n", __FUNCTION__, keybits);
-	  dumpHex((uint8_t *)key,keybits/8);
-	
+
     /* Also checks keybits */
     if( ( ret = mbedtls_aes_setkey_enc( ctx, key, keybits ) ) != 0 )
-        goto exit;    
+        goto exit;
 
 exit:
-
     return( ret );
 }
-#endif /* MBEDTLS_AES_SETKEY_DEC_ALT */
-
 
+/* Do AES encrypt/decrypt with H/W accelerator
+ *
+ * NOTE: As input/output buffer doesn't follow constraint of DMA buffer, static allocated 
+ *       DMA compatible buffer is used for DMA instead and this needs extra copy.
+ *
+ * NOTE: dataSize requires to be: 
+ *       1) Multiple of block size 16 
+ *       2) <= MAX_DMA_CHAIN_SIZE
+ */
 static void __nvt_aes_crypt( mbedtls_aes_context *ctx,
-                          const unsigned char input[16],
-                          unsigned char output[16], int dataSize)
+                             const unsigned char *input,
+                             unsigned char *output, size_t dataSize)
 {
-		unsigned char* pIn;
-	  unsigned char* pOut;
-
-//	  mbedtls_trace("=== %s \r\n", __FUNCTION__);
-	  dumpHex(input,16);
- 
-    AES_Open(ctx->channel, ctx->encDec, ctx->opMode, ctx->keySize, ctx->swapType);
-    AES_SetInitVect(ctx->channel, ctx->iv);
-		if( ((uint32_t)input) & 0x03 )
-		{
-			memcpy(au8InputData, input, dataSize);
-			pIn = au8InputData;
-		}else{
-		  pIn = (unsigned char*)input;
+    const unsigned char* pIn;
+    unsigned char* pOut;
+
+    MBED_ASSERT((dataSize % 16 == 0) && (dataSize <= MAX_DMA_CHAIN_SIZE));
+    
+    /* AES DMA buffer requires to be:
+     *   1) Word-aligned
+     *   2) Located in 0x2xxxxxxx region
+     */
+    if ((! crypto_dma_buff_compat(au8OutputData, MAX_DMA_CHAIN_SIZE)) || (! crypto_dma_buff_compat(au8InputData, MAX_DMA_CHAIN_SIZE))) {
```

Comment:
Style: There is a lot of unnecessary brackiting here. I would suggest the following cosmetic change here for readability:
```C
if (!crypto_dma_buff_compat(au8OutputData, MAX_DMA_CHAIN_SIZE) ||
    !crypto_dma_buff_compat(au8InputData, MAX_DMA_CHAIN_SIZE)) {
```

Reply:
@andresag01 I keep the parentheses to avoid operator precedence confusion.